DataONEorg/slinky
# Port the essential services from Docker to Kubernetes

This PR is mainly concerned with porting the essentials over to Kubernetes.

Fixes #3 
Fixes #10 

### Changes:
1. The first change is that I removed all of the docker related files.
2. Updated the README to reflect the Kubernetes deployment
3. Added a new directory, `deploy/` that contains the K8 Helm Chart
4. Refactored the Docker Image builds so that the commands are now in the K8 config files
5. Removed the `data/` folder since K8 works with PersistentVolume objects

### Kubernetes Details:
The most controversial thing that I can think of is that copying files from the host into a pod isn't a Kubernetes pattern. This amounts to some refactoring done around the Docker [COPY](https://github.com/DataONEorg/slinky/blob/main/scheduler/Dockerfile#L11) directive. The best way to do this is to bundle the requirements into a docker image, hosted on dockerhub, and pulled by K8. This allows for a streamlined deployment process that's agnostic of the hosts' filesystem. For the moment, I'm using my personal dockerhub to hold the scheduler & worker images. These should be moved over to the DataONE dockerhub at some point.

K8 still has a 'master' role for services, it looks like the issue is currently [unresolved](https://github.com/kubernetes/website/issues/6525) (it's apparently easier to close an issue that sparks debate than actually deal with it). I'll keep my eyes open for ways to come up with a better name.

### To Test:

1. Install & start minikube
2. Clone this repository and navigate to `deploy/`
3. Run `helm install ./ --generate-name`
4. Check that the pods and services are up
5. Run `minikube service virtuoso`
6. See that the virtuoso UI opens
7. Run a sparql query to check that data from DataONE has been inserted

#### Bonus Testing
Play around with setting replicas to 0 for various pods and check the logs of other pods to make sure they fail. Then bring the replica count back up and see that the other pods successfully find the pod. For example setting the replica count for the worker pod to 0 will cause the scheduler to fail. After setting it back to something greater than or equal to 1, the worker should be able to find it and continue processing data.


### Reviewing
The commits are fairly isolated and can be reviewed individually. 456bb45 concerns Kubernetes management and 683d625 concerns adding the Dockerfiles so that the Kubernetes stack can be updated with code changes.
